### PR TITLE
No repeated primes in the limb moduli for CKKS

### PR DESCRIPTION
**Pull request by dsuponitskiy**
1. Make sure that the vector of moduliQ in ParamsGenCKKSRNS() has no duplicate values.
2. Push error handling to the top of the function
3. Reorganize code and replace usint with uint32_t
4. Add unit tests

### CODE REVIEW

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 1]
```diff
@@ -122,107 +128,119 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(std::shared_ptr<CryptoParamete
...
 
                 rootsQ[i] = RootOfUnity(cyclOrder, moduliQ[i]);
-                cnt++;
             }
         }
     }
 
     if (firstModSize == dcrtBits) {  // this requires dcrtBits < 60
-        moduliQ[0] = PreviousPrime<NativeInteger>(qPrev, cyclOrder);
+        moduliQ[0] = PreviousPrime<NativeInteger>(minPrime, cyclOrder);
```
Comment (andreea-alexandru) [inline src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:209]: Flag to double check the logic here.

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 2]
```diff
@@ -122,107 +128,119 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(std::shared_ptr<CryptoParamete
...
+        moduliQ[0] = NextPrime<NativeInteger>(maxPrime, cyclOrder);
+        maxPrime   = moduliQ[0];
+    }
     rootsQ[0] = RootOfUnity(cyclOrder, moduliQ[0]);
 
     if (scalTech == FLEXIBLEAUTOEXT) {
         // no need for extra checking as extraModSize is automatically chosen by the library
-        moduliQ[numPrimes] = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
-        rootsQ[numPrimes]  = RootOfUnity(cyclOrder, moduliQ[numPrimes]);
+        NativeInteger temp_moduliQ = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
```
Comment (andreea-alexandru) [inline src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:225]: Why not do this inside the if branch?

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp (hunk 1 above)
Comment (andreea-alexandru) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:209]: Should be moduliQ[0] = NextPrime<NativeInteger>(maxPrime, cyclOrder);

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 3]
```diff
@@ -122,107 +128,119 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(std::shared_ptr<CryptoParamete
...
-        moduliQ[0] = PreviousPrime<NativeInteger>(qPrev, cyclOrder);
+        moduliQ[0] = PreviousPrime<NativeInteger>(minPrime, cyclOrder);
     }
     else {
         moduliQ[0] = LastPrime<NativeInteger>(firstModSize, cyclOrder);
     }
+    // find if the value of moduliQ[0] is already in the vector starting with moduliQ[1] and
+    // if there is, then get another prime for moduliQ[0]
+    const auto pos = std::find(moduliQ.begin() + 1, moduliQ.end(), moduliQ[0]);
+    if (pos != moduliQ.end()) {
```
Comment (andreea-alexandru) [inline src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:217]: This check should be inside the else branch  from above.

**Review by andreea-alexandru — COMMENTED** [review]
Please add a unit test checking for small firstModSize = scalingModSize and for small firstModSize = scalingModSize + 1.

@dsuponitskiy You should be able to reuse the examples from https://openfhe.discourse.group/t/crashes-with-specific-parameter-setting-in-ckks/1252/4

**Review by andreea-alexandru — APPROVED** [review]
Looks good to me.

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 4]
```diff
@@ -125,107 +131,129 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(std::shared_ptr<CryptoParamete
...
 
     if (scalTech == FLEXIBLEAUTOEXT) {
-        // no need for extra checking as extraModSize is automatically chosen by the library
-        moduliQ[numPrimes] = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
-        rootsQ[numPrimes]  = RootOfUnity(cyclOrder, moduliQ[numPrimes]);
+        // find if the value of moduliQ[numPrimes] is already in the vector.
+        // in this case we must redefine the position of "end()" as "end()-1"
+        // currently "moduliQ[numPrimes] == 0" as it is not set up anywhere in this code yet
+        const auto endPos = moduliQ.end() - 1;
+        auto pos          = std::find(moduliQ.begin(), endPos, moduliQ[numPrimes]);
```
Comment (yspolyakov) [inline src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:234]: I don't understand what this line does. Why are we looking for a value = 0 = moduliQ[numPrimes] in the vector of modulus between 0 and numPrimes - 1?

Comment (dsuponitskiy) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:234]: will change and add extra comments

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 5]
```diff
@@ -125,107 +131,122 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(std::shared_ptr<CryptoParamete
...
     rootsQ[0] = RootOfUnity(cyclOrder, moduliQ[0]);
 
     if (scalTech == FLEXIBLEAUTOEXT) {
+        // moduliQ[numPrimes] must still be 0, so it has to be populated now
+
         // no need for extra checking as extraModSize is automatically chosen by the library
-        moduliQ[numPrimes] = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
-        rootsQ[numPrimes]  = RootOfUnity(cyclOrder, moduliQ[numPrimes]);
+        auto tempMod = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
+        if (tempMod > maxPrime)
```
Comment (yspolyakov) [inline src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:234]: Why is this code needed? tempMod can be larger than maxPrime only if it is not in moduliQ. Yet, you use maxPrime (below) only when tempMod is in moduliQ. So the branch and re-assignment of maxtPrime to tempMod is the code that is never executed.

**Review by yspolyakov — APPROVED** [review]

**Pull request merged**